### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by djmitche**
This is, I think, a complete initial implementation that supports connecting and reconnecting.  I'll make further PRs to start incorporating the functionality of taskcluster-client and pulse-publisher.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+/**
+ * A single connection to a pulse server.  This is a thin wrapper around a raw
+ * AMQP connection, instrumented to inform the parent Client of failures
+ * and trigger a reconnection.  It is possible to have multiple Connection
+ * objects in the same process at the same time, while one is being "retired" but
+ * is lingering around to send ack's for any in-flight message handlers.
+ *
+ * The instance's `amqp` property is the amqp connection object.  In the event of any
+ * issues with the connection, call the instance's `failed` method.  This will initiate
```
Comment (jonasfj) [inline src/index.js:148]: Why don't we detect by listening to the `error` event... calling failed seems like a bad idea...

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        // ignore..
+      }
+      this.amqp = null;
+      this.state = 'finished';
+      this.emit('finished');
+    }, this.client.retirementDelay);
+  }
+}
+
+exports.Connection = Connection;
```
Comment (jonasfj) [inline src/index.js:264]: Can we avoid exporting this? since this should only be used within this package?

**Review by jonasfj — APPROVED** [review]
few changes, but otherwise r+

File: src/index.js (hunk 2 above)
Comment (djmitche) [reply src/index.js:264]: It's exported for tests.. if only there was a `pub(crate)` in JS :)

File: src/index.js (hunk 1 above)
Comment (djmitche) [reply src/index.js:148]: We do that, too.  This is more of a belt-and-suspenders approach, where if the user sees some issue on the connection (like a timeout) they can also flag it as a failure.

**Pull request merged**